Implement scoring for extended resources backed up by DRA

bart0sh opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Currently, the scheduler only scores extended resources provided via device plugins. Extended resources backed by DRA are ignored in scoring, which makes the behavior inconsistent with the existing scheduling model:

For device-plugin backed extended resources, allocatable and requested values are included in node scoring.

For DRA-backed extended resources, the same requests are currently excluded.

This effectively breaks backward compatibility for workloads that migrate from device-plugin extended resources to DRA-backed ones. Instead of preserving consistent placement decisions, the scheduler would treat these resources as if they don’t matter at all, leading to incompatible scheduling decisions.

This gap must be closed for DRA extended resources to graduate to Beta.

#### Which issue(s) this PR is related to:

Fixes: https://github.com/kubernetes/kubernetes/issues/133669

#### Special notes for your reviewer:

##### High level technical details of the implementation:

The implementation calculates total number of available devices (capacity/allocatable) and requested(already allocated + requested by pod) amount of devices for extended resources backed by DRA for a node. For backward compatibility, device plugin resources are still handled first; DRA-backed extended resources are considered only when `Allocatable` value for extended resource is 0. 

The computed (capacity, allocated + podRequest) values are fed back into the existing scoring logic, just like for CPU, memory, or device plugin resources.

`SharedDraManager` is used to get information about device classes, resource slices and resource claims to simplify the implementation and avoid duplicating dynamicresources APIs.

Capacity is calculated by counting devices that belong to the resource slices on the node that match the resource device class. Allocated is calculated by counting already allocated devices returned by dra manager ResourceClaims().ListAllAllocatedDevices() API and filtering out devices from other nodes and device classes.

Device filtering honors node selectors and CEL-based match expressions from the DeviceClass.

#### Does this PR introduce a user-facing change?

```release-note
Implement scoring for DRA-backed extended resources
```

bart0sh committed:
DRA: implement scoring for extended resources

Updated extended resource allocation scorer to calculate
allocatable and requested values for DRA-backed resources.

bart0sh commented:
/test pull-kubernetes-unit
/test pull-kubernetes-verify
/test pull-kubernetes-typecheck
/test pull-kubernetes-dra-all

pohly commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +171,142 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+		// Count devices that match the device class
+		for _, device := range slice.Spec.Devices {
+			if r.deviceMatchesClass(ctx, device, deviceClass, driver) {
+				totalCapacity++
```
If [`slice.spec.perDeviceNodeSelection`](https://github.com/kubernetes/kubernetes/blob/74c336734c0eda74f6e9ae9bc82c7470bec7640f/staging/src/k8s.io/api/resource/v1/types.go#L157-L167) is true, then you need to check for each device whether it is available for the node. This needs another unit test case.

bart0sh replied:
Done

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -193,3 +193,8 @@ var availableAllocators = []struct {
 	},
 }
+
+// NodeMatches wraps stable.NodeMatches from internal/stable to make it available to other packages.
```
Hmm, this assumes that the implementation is the same across all implementations. What if features are active where e.g. `experimental.NodeMatches` behaves differently? See the comment about `PerDeviceNodeSelection`...

bart0sh commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -105,11 +112,20 @@ func (r *resourceAllocationScorer) calculateResourceAllocatableRequest(logger kl
+		if allocatable == 0 && r.enableDRAExtendedResource {
+			// Allocatable 0 means that this resource is not handled by device plugin.
+			// Calculate allocatable and requested for resources backed by DRA.
+			allocatable, requested := r.calculateDRAExtendedResourceAllocatableRequest(ctx, nodeInfo.Node(), resource, podRequest)
```
The code is called only for chosen nodes, the same way as existing scoring code for extended resources. This ensures backward compatibility, which is a main goal of this implementation.

bart0sh commented:
/cc @johnbelamaric

bart0sh commented:
/skip
/retest

bart0sh commented:
/test pull-kubernetes-integration
/skip

bart0sh commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +173,186 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+		}
+		// Count devices that match the device class
+		for _, device := range devices {
+			if r.deviceMatchesClass(ctx, device, deviceClass, driver) {
```
As we filter out devices backed up only by DRA, this situation shouldn't occur. We have 1:1 resource-class mapping, so one device can't match more than one class in my understanding. Am I missing something? Can you give an example?

bart0sh replied:
Which of those 2 types of resources will the pod request in this case?
In my understanding it should request the one that suits both containers: gpu.example.com/device: 2, which would map the first class. If it requests two different resources gpu.example.com/device: 1 and gpu-1g-20gb.example.com: 1, they would be scored separately, and the mapping will still be 1:1, no?

yliaog replied:
a device of gpu-1g-20gb.example.com is also counted in gpu.example.com/device, this is different from the extended resources backed by device plugin, which are distinct from each other.

bart0sh replied:
At this level we only calculate total capacity of one resource(let's say for gpu.example.com). We don't know if the same pod requested gpu-1g-20gb.example.com or not and how many. I don't think we should consider this test case as It would require caching of requested resources and their classes. In reality I'm not sure anybody would use this setup as workload would have to somehow find out which device to use as they're both available for all pod containers.

bart0sh replied:
@yliaog Correct me if I'm wrong here, you're suggesting to consider pod requests when calculating total node capacity for certain device, correct? So, if pod requests gpu.example.com/device and  gpu-1g-20gb.example.com device, total capacity should be different than if pod requests two gpu.example.com/device or two gpu-1g-20gb.example.com/device ? Can you explain how exactly it should be different?

bart0sh commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +171,142 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+				totalCapacity++
+				// Count allocated devices
+				deviceID := structured.MakeDeviceID(driver, pool, device.Name)
+				if allocatedDevices.Has(deviceID) {
```
Thank you for pointing out! done.

bart0sh commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +173,186 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+		result := r.celCache.GetOrCompile(selector.CEL.Expression)
+		if result.Error != nil {
+			// If compilation fails, consider it a non-match
+			continue
```
Fixed, thanks for pointing out. It's even not a non-match, but an error.

yliaog commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +173,186 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+	// Only consider device classes that have ExtendedResourceName set
+	// This ensures we only score DRA resources that are meant to be compatible
+	// with device plugin extended resources
+	if ptr.Deref(deviceClass.Spec.ExtendedResourceName, "") != string(resource) {
```
that's ok. but it seems there is no need to differentiate between extended resources backed by device plugin, then switching to DRA devices, and those backed only by DRA, is that correct?

bart0sh replied:
I'm ok with removing this condition. I thought that it may conflict with the general DRA scoring in future, so it's better to be at the safe place. No strong opinion with this. I can remove if you think it's better.

yliaog replied:
General DRA scoring has to consider this, and make necessary changes. I think it is helpful to allow for implicit extended resource in this PR. Because for Beta, implicit extended resource needs to be supported, and it should have the same scoring support as regular extended resource.

yliaog commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +173,186 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+// Note: This method assumes the device class has ExtendedResourceName set, as filtering
+// should be done by the caller to ensure we only process DRA resources meant for extended
+// resource scoring.
+func (r *resourceAllocationScorer) deviceMatchesClass(ctx context.Context, device resourceapi.Device, deviceClass *resourceapi.DeviceClass, driver string) bool {
```
I'm not sure, just stating in general it seems to a general DRA util that could be used elsewhere. I'll leave it up to you.

bart0sh commented:
/retest

macsko commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +173,152 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+// calculateDRAExtendedResourceAllocatableRequest calculates allocatable and requested
+// quantities for extended resources backed by DRA.
+func (r *resourceAllocationScorer) calculateDRAExtendedResourceAllocatableRequest(ctx context.Context, node *v1.Node, resource v1.ResourceName, podRequest int64) (int64, int64) {
+	logger := klog.FromContext(ctx).WithValues("resource", resource, "node", node.Name)
```
Please set values only when we need. For example:
```suggestion
	logger := klog.FromContext(ctx)
	if logger.V(5).Enabled() {
		logger = logger.WithValues("resource", resource, "node", node.Name)
	}
```

bart0sh replied:
After using logger.Error instead of `logger.Info` (see your comments below) only one `logger.V(7).Info` call is left, so I think it's not needed anymore. Let me know if you still think we need this.

macsko commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +173,152 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+	// Get device class mapping to find the device class for this resource
+	deviceClassMapping, err := extended.DeviceClassMapping(r.draManager)
+	if err != nil {
+		logger.V(5).Info("Failed to get device class mapping for DRA extended resource scoring", "error", err)
```
Why are we logging that errors as Info?

bart0sh replied:
Fixed, thanks for pointing out!

macsko reviewed (commented):
Initial comments

bart0sh committed:
Cache DRA state for scoring extended resources

Extend Fit and BalancedAllocation PreScore state with the the
allocated state, the list of ResourceSlices and the device class
mapping. Gather these once during PreScore and pass them through
the scoring path instead of re-fetching for every scoring call.

This should speed up scoring of DRA extended resources, lowering
scheduling overhead.

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>
Co-authored-by: Maciej Skoczeń <mskoczen@google.com>
Co-authored-by: Dominik Marciński <gmidon@gmail.com>

bart0sh committed:
DRA: test performance for scoring extended resources

macsko commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
How does the benchmark look like? Could you share the test case?

bart0sh replied:
It's included into this PR: https://github.com/kubernetes/kubernetes/pull/134058/files#diff-62183a3486d02ad39f70b9400d3b9308db2ec35b297e6c65f4c7dd919d78b6e1

bart0sh replied:
Here is a pprof output sorted by cumulative % for the master branch and this pr. It shows that most of the performance impact (1066s or 37.99%) brought by the runtime and only 251s or 8.95% - by the new code introduced by this PR. I don't fully understand the reason yet and will continue digging into it. I'd appreciate any ideas and suggestions on this.

**Pprof output**

**master:**
<img width="1551" height="862" alt="master" src="https://github.com/user-attachments/assets/0d4e448b-ce25-4c8c-a5db-d71919da0cfe" />


**this pr:**

<img width="1648" height="866" alt="pr" src="https://github.com/user-attachments/assets/9323e42f-2310-4a6c-b629-0d36f675f231" />

bart0sh replied:
> what is the number for regular extended resources with and without scoring enabled? i think we want to compare that with DRA extended resources with and without scoring.

I'm not sure it will give us anything different. Scoring of extended resources (without DRA) is a way simpler than this implementation. It has all numbers at hand: allocatable is NodeInfo.Allocatable, and requested is NodeInfo.Requested. No heavy calculations or CEL expression evealuation, no loops over slices and devices etc. 
So, comparing with DRA scoring time, time of non-DRA scoring is around 0.

Here is how a performance picture of the DRA scoring looks like:

<img width="2373" height="4181" alt="cpu_with_scoring" src="https://github.com/user-attachments/assets/be4a6436-a6bb-4b8c-8e14-1088ec7aa0be" />

Main consumers are `structured.NodeMatches` and `deviceMatchesClass`. I'm trying to cache their results, will see how it goes. In any case the results will be absolutely incomparable with non-DRA scoring in my opinion.

bart0sh replied:
@macsko Caching NodeMatches and deviceMatches results improved average throughput from 125.5 to 210.5 pods/sec. PTAL. I'm still hoping that this PR can be merged before the code freeze.

Here is a full benchstat output:
```
| a                                                           | master.log                                              | c   | scoring.log                                             | e     | f       | g            |
| ----------------------------------------------------------- | ------------------------------------------------------- | --- | ------------------------------------------------------- | ----- | ------- | ------------ |
|                                                             | SchedulingThroughput/Average                            | CI  | SchedulingThroughput/Average                            | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 362.5                                                   | 2%  | 210.5                                                   | 12%   | -41.93% | p=0.000 n=10 |
| geomean                                                     | 362.49999999999994                                      |     | 210.49999999999991                                      |       | -41.93% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | SchedulingThroughput/Perc50                             | CI  | SchedulingThroughput/Perc50                             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 294.55                                                  | 7%  | 173.5                                                   | 3%    | -41.10% | p=0.000 n=10 |
| geomean                                                     | 294.55000000000007                                      |     | 173.5                                                   |       | -41.10% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | SchedulingThroughput/Perc90                             | CI  | SchedulingThroughput/Perc90                             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 536.05                                                  | 2%  | 347.20000000000005                                      | 2%    | -35.23% | p=0.001 n=10 |
| geomean                                                     | 536.0499999999998                                       |     | 347.2000000000001                                       |       | -35.23% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | SchedulingThroughput/Perc95                             | CI  | SchedulingThroughput/Perc95                             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 536.05                                                  | 2%  | 349.8                                                   | 3%    | -34.74% | p=0.001 n=10 |
| geomean                                                     | 536.0499999999998                                       |     | 349.8                                                   |       | -34.74% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | SchedulingThroughput/Perc99                             | CI  | SchedulingThroughput/Perc99                             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 536.05                                                  | 2%  | 349.8                                                   | 3%    | -34.74% | p=0.001 n=10 |
| geomean                                                     | 536.0499999999998                                       |     | 349.8                                                   |       | -34.74% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | heap_memory_usage/Average                               | CI  | heap_memory_usage/Average                               | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 522.65                                                  | 3%  | 548.35                                                  | 3%    | +4.92%  | p=0.002 n=10 |
| geomean                                                     | 522.6500000000001                                       |     | 548.3499999999998                                       |       | +4.92%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | heap_memory_usage/Max                                   | CI  | heap_memory_usage/Max                                   | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 590.75                                                  | 3%  | 629.25                                                  | 18%   | +6.52%  | p=0.002 n=10 |
| geomean                                                     | 590.7499999999998                                       |     | 629.2499999999999                                       |       | +6.52%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | heap_memory_usage/Perc50                                | CI  | heap_memory_usage/Perc50                                | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 523.2                                                   | 4%  | 550.95                                                  | 4%    | +5.30%  | p=0.003 n=10 |
| geomean                                                     | 523.2                                                   |     | 550.9500000000002                                       |       | +5.30%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | heap_memory_usage/Perc90                                | CI  | heap_memory_usage/Perc90                                | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 573.4                                                   | 1%  | 597.65                                                  | 2%    | +4.23%  | p=0.000 n=10 |
| geomean                                                     | 573.3999999999999                                       |     | 597.65                                                  |       | +4.23%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | heap_memory_usage/Perc95                                | CI  | heap_memory_usage/Perc95                                | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 590.75                                                  | 3%  | 607.85                                                  | 5%    | +2.89%  | p=0.015 n=10 |
| geomean                                                     | 590.7499999999998                                       |     | 607.85                                                  |       | +2.89%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | heap_memory_usage/Perc99                                | CI  | heap_memory_usage/Perc99                                | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 590.75                                                  | 3%  | 629.25                                                  | 18%   | +6.52%  | p=0.002 n=10 |
| geomean                                                     | 590.7499999999998                                       |     | 629.2499999999999                                       |       | +6.52%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | memory_growth_rate/GrowthRate                           | CI  | memory_growth_rate/GrowthRate                           | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 1758                                                    | 18% | 1118                                                    | 23%   | -36.41% | p=0.002 n=10 |
| geomean                                                     | 1757.9999999999998                                      |     | 1118                                                    |       | -36.41% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | runtime_seconds                                         | CI  | runtime_seconds                                         | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 27.46                                                   | 2%  | 31.495                                                  | 2%    | +14.69% | p=0.000 n=10 |
| geomean                                                     | 27.46                                                   |     | 31.495                                                  |       | +14.69% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_event_handling_duration_ms/Average            | CI  | scheduler_event_handling_duration_ms/Average            | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.01091                                                 | 10% | 0.010675                                                | 23%   | ~       | p=0.912 n=10 |
| geomean                                                     | 0.010909999999999994                                    |     | 0.010675000000000002                                    |       | -2.15%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_event_handling_duration_ms/Perc50             | CI  | scheduler_event_handling_duration_ms/Perc50             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.05                                                    | 0%  | 0.05                                                    | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.05                                                    |     | 0.05                                                    |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_event_handling_duration_ms/Perc90             | CI  | scheduler_event_handling_duration_ms/Perc90             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.09                                                    | 0%  | 0.09                                                    | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.09000000000000001                                     |     | 0.09000000000000001                                     |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_event_handling_duration_ms/Perc95             | CI  | scheduler_event_handling_duration_ms/Perc95             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.095                                                   | 0%  | 0.095                                                   | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.09500000000000001                                     |     | 0.09500000000000001                                     |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_event_handling_duration_ms/Perc99             | CI  | scheduler_event_handling_duration_ms/Perc99             | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.099                                                   | 0%  | 0.099                                                   | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.099                                                   |     | 0.099                                                   |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_framework_extension_point_duration_ms/Average | CI  | scheduler_framework_extension_point_duration_ms/Average | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.0009410499999999999                                   | 8%  | 0.0010495                                               | 9%    | +11.52% | p=0.000 n=10 |
| geomean                                                     | 0.0009410500000000002                                   |     | 0.0010495000000000005                                   |       | +11.52% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_framework_extension_point_duration_ms/Perc50  | CI  | scheduler_framework_extension_point_duration_ms/Perc50  | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.05                                                    | 0%  | 0.05                                                    | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.05                                                    |     | 0.05                                                    |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_framework_extension_point_duration_ms/Perc90  | CI  | scheduler_framework_extension_point_duration_ms/Perc90  | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.09                                                    | 0%  | 0.09                                                    | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.09000000000000001                                     |     | 0.09000000000000001                                     |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_framework_extension_point_duration_ms/Perc95  | CI  | scheduler_framework_extension_point_duration_ms/Perc95  | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.095                                                   | 0%  | 0.095                                                   | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.09500000000000001                                     |     | 0.09500000000000001                                     |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_framework_extension_point_duration_ms/Perc99  | CI  | scheduler_framework_extension_point_duration_ms/Perc99  | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.099                                                   | 0%  | 0.099                                                   | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.099                                                   |     | 0.099                                                   |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_plugin_execution_duration_ms/Average          | CI  | scheduler_plugin_execution_duration_ms/Average          | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.00068995                                              | 24% | 0.0008101499999999999                                   | 16%   | +17.42% | p=0.002 n=10 |
| geomean                                                     | 0.0006899499999999999                                   |     | 0.0008101499999999996                                   |       | +17.42% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_plugin_execution_duration_ms/Perc50           | CI  | scheduler_plugin_execution_duration_ms/Perc50           | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.005                                                   | 0%  | 0.005                                                   | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.005000000000000002                                    |     | 0.005000000000000002                                    |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_plugin_execution_duration_ms/Perc90           | CI  | scheduler_plugin_execution_duration_ms/Perc90           | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.009                                                   | 0%  | 0.009                                                   | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.009                                                   |     | 0.009                                                   |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_plugin_execution_duration_ms/Perc95           | CI  | scheduler_plugin_execution_duration_ms/Perc95           | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.0095                                                  | 0%  | 0.0095                                                  | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.0095                                                  |     | 0.0095                                                  |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_plugin_execution_duration_ms/Perc99           | CI  | scheduler_plugin_execution_duration_ms/Perc99           | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 0.0099                                                  | 0%  | 0.0099                                                  | 0%    | ~       | p=1.000 n=10 |
| geomean                                                     | 0.009899999999999999                                    |     | 0.009899999999999999                                    |       | +0.00%  |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_scheduling_attempt_duration_ms/Average        | CI  | scheduler_scheduling_attempt_duration_ms/Average        | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 43.965                                                  | 39% | 33.94                                                   | 1098% | ~       | p=0.247 n=10 |
| geomean                                                     | 43.965                                                  |     | 33.94                                                   |       | -22.80% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_scheduling_attempt_duration_ms/Perc50         | CI  | scheduler_scheduling_attempt_duration_ms/Perc50         | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 32.095                                                  | 51% | 35.88                                                   | 54%   | ~       | p=0.382 n=10 |
| geomean                                                     | 32.095                                                  |     | 35.88                                                   |       | +11.79% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_scheduling_attempt_duration_ms/Perc90         | CI  | scheduler_scheduling_attempt_duration_ms/Perc90         | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 62.495000000000005                                      | 19% | 54.295                                                  | 2979% | ~       | p=0.190 n=10 |
| geomean                                                     | 62.494999999999976                                      |     | 54.29500000000001                                       |       | -13.12% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_scheduling_attempt_duration_ms/Perc95         | CI  | scheduler_scheduling_attempt_duration_ms/Perc95         | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 73.66499999999999                                       | 49% | 59.325                                                  | 3694% | ~       | p=0.190 n=10 |
| geomean                                                     | 73.665                                                  |     | 59.32500000000002                                       |       | -19.47% |              |
|                                                             |                                                         |     |                                                         |       |         |              |
|                                                             | master.log                                              |     | scoring.log                                             |       |         |              |
|                                                             | scheduler_scheduling_attempt_duration_ms/Perc99         | CI  | scheduler_scheduling_attempt_duration_ms/Perc99         | CI    | vs base | P            |
| PerfScheduling/ExtendedResourceScoring/2000pods_200nodes-24 | 117.15                                                  | 75% | 75.59                                                   | 4831% | ~       | p=0.481 n=10 |
| geomean                                                     | 117.15000000000002                                      |     | 75.59000000000003                                       |       | -35.48% |              |
```

bart0sh commented:
/retest

macsko commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -94,7 +98,7 @@ type Fit struct {
 	enableDRAExtendedResource       bool
 	handle                          fwk.Handle
-	resourceAllocationScorer
+	*resourceAllocationScorer
```
Does this change bring any value?

macsko commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -37,19 +44,99 @@ type resourceAllocationScorer struct {
+	// Cache for DeviceMatches results to avoid expensive repeated evaluations
+	deviceMatchCache sync.Map // map[deviceMatchCacheKey]bool
+	// Cache for NodeMatches results to avoid expensive repeated node selector evaluations
+	nodeMatchCache sync.Map // map[nodeMatchCacheKey]bool
```
Maybe move it to a separate struct as well (and link it here)?

bart0sh replied:
Done

macsko reviewed (commented):
Not reviewing fully, as it needs a rebase for extended resource cache

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:
```diff
@@ -103,14 +126,23 @@ func (ba *BalancedAllocation) Score(ctx context.Context, state fwk.CycleState, p
 			return 0, nil
 		}
+		if ba.enableDRAExtendedResource {
+			allocatedState, resourceSlices, deviceClassMapping, status := getDRAPreScoredParams(ba.draManager, ba.resources)
```
We already have it in the state form PreScore, right?

dom4ha replied:
Hmm, it seems to be an existing pattern to repeat prescoring on error, although I don't get why.

@sanposhiho @macsko

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -661,7 +718,16 @@ func (f *Fit) Score(ctx context.Context, state fwk.CycleState, pod *v1.Pod, node
 		s = &preScoreState{
 			podRequests: f.calculatePodResourceRequestList(pod, f.resources),
 		}
+		if f.enableDRAExtendedResource {
```
Check additionally whether canBeHandledByDRA

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -124,11 +128,54 @@ func (s *preScoreState) Clone() fwk.StateData {
+		}
+
+		// check if resource is mapped to device class (can be handled by DRA)
+		if f.draManager.DeviceClassResolver().GetDeviceClass(resource) == "" {
```
What about moving the content of `canBeHandledByDRA` to `getDRAPreScoredParams` once we remove iterating over nodes? This way it would be always called even for the balanced allocation plugin

This way `getDRAPreScoredParams` could return early empty results if there is no extended resource in device class mapping, which `getDRAPreScoredParams` needs to construct anyway.

bart0sh replied:
@dom4ha Thank you for the review. Done, PTAL.

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -32,24 +38,111 @@ import (
+func buildDeviceMatchCacheKey(expression string, driver string, poolName string, deviceName string) string {
+	// Pre-allocate sufficient capacity to avoid reallocation
+	var b strings.Builder
+	b.Grow(len(expression) + len(driver) + len(poolName) + len(deviceName) + 4)
```
```suggestion
	b.Grow(len(expression) + len(driver) + len(poolName) + len(deviceName) + 3)
```

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +265,225 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+			return nil, nil, nil, fwk.AsStatus(err)
+		}
+		deviceClassMapping[resourceName] = deviceClass
+	}
```
IIUC you can return early if len(deviceClassMapping) == 0

bart0sh commented:
@dom4ha All review comments have been addressed. PTAL again, thanks!

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -222,5 +229,49 @@ var availableAllocators = []struct {
+	}
+
+	// Check if device is partially consumed via shared allocations (consumable capacity case)
+	// We need to check if any shared device ID corresponds to our device
```
```suggestion
	// Check if device is partially consumed via shared allocations (consumable capacity case).
	// We need to check if any shared device ID corresponds to our device.
```

My usual nit :grin: - lack of punctuation. I also prefer to have it elsewhere because one never knows when another (partial) sentence may have to be added.

But this is minor and I am not calling it out elsewhere.

pohly reviewed (commented):
Some nits. I can't vouch for the correctness of the code (in particular in the scheduler) but am okay with going ahead.

/approve

For DRA. Final review/approval through SIG Scheduling.

pohly commented:
/milestone v1.35

Assuming that we can satisfy reviewers.

dom4ha reviewed (commented):
Looks good now. Leaving approval on @macsko who is back and wants to have the last pass.

macsko reviewed (commented):
Change looks correct, left some minor comments

bart0sh commented:
@dom4ha @pohly @macsko Thank you for the review! All your review comments should be addressed now, PTAL!

macsko commented on pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:
```diff
@@ -73,6 +85,15 @@ func (ba *BalancedAllocation) PreScore(ctx context.Context, cycleState fwk.Cycle
 	state := &balancedAllocationPreScoreState{
 		podRequests: podRequests,
 	}
+	if ba.enableDRAExtendedResource && state.draPreScoreState == nil {
```
`state.draPreScoreState` is always nil here
```suggestion
	if ba.enableDRAExtendedResource {
```

macsko commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
```diff
@@ -155,3 +261,231 @@ func (r *resourceAllocationScorer) isBestEffortPod(podRequests []int64) bool {
+	deviceClass := r.draManager.DeviceClassResolver().GetDeviceClass(resource)
+	if deviceClass == nil {
+		// This resource is not backed by DRA.
+		logger.Error(nil, "Extended resource not found in device class mapping", "resource", resource)
```
Shouldn't this be logged as info?

bart0sh replied:
fixed, thanks!

yliaog commented:
/retest

macsko commented:
/approve
Scheduling

bart0sh commented on pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:
```diff
@@ -103,14 +124,23 @@ func (ba *BalancedAllocation) Score(ctx context.Context, state fwk.CycleState, p
 		if ba.isBestEffortPod(s.podRequests) {
 			return 0, nil
 		}
+		if ba.enableDRAExtendedResource && s.draPreScoreState == nil {
```
done, thanks!

macsko commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -701,7 +726,16 @@ func (f *Fit) Score(ctx context.Context, state fwk.CycleState, pod *v1.Pod, node
 		s = &preScoreState{
 			podRequests: f.calculatePodResourceRequestList(pod, f.resources),
 		}
+		if f.enableDRAExtendedResource && s.draPreScoreState == nil {
```
Similarly: https://github.com/kubernetes/kubernetes/pull/134058#discussion_r2499648377

bart0sh commented:
/retest
